### PR TITLE
feat: add maui port check command

### LINKED ISSUE #197: DevFlow skills: prefer unified maui CLI over standalone tools + improve JSON/error guidance
## Summary

The `maui-devflow-debug` and `maui-devflow-onboard` agent skills (under `plugins/dotnet-maui/skills/`) currently steer AI agents toward a mix of clients to prepare and handle devices:

- `androidsdk.tool` (`android sdk|avd|device|jdk …`)
- `appledev.tools` (`apple xcode|runtime|simulator …`)
- Raw `adb …` and `xcrun simctl …`
- Plus `lsof`, `pgrep`, etc.

Most of those operations are **already covered by the unified `maui` CLI** (`Microsoft.Maui.Cli`). Mixing tools forces agents to install three global tools, learn three command surfaces, and parse three different output shapes. JSON / error guidance in the skills is also weak — agents are not told to use `--json`, parse `error.code`, or follow `remediation.command` hints that the CLI already emits via `MauiToolException`.

This issue tracks **(a)** retargeting the active DevFlow skills to the `maui` CLI everywhere it supports the operation today, **(b)** improving JSON and error-troubleshooting guidance inline, and **(c)** capturing the real `maui` CLI gaps as follow-up work.

## Audit — `maui` CLI coverage matrix

Verified against `src/Cli/Microsoft.Maui.Cli/Commands/`.

| Skill operation | Old reference | `maui` equivalent | Status |
|---|---|---|---|
| List devices/emulators | `adb devices` / `xcrun simctl list devices booted` / `apple simulator list` / `android device list` | `maui device list --platform android\|apple` | ✅ replace |
| List AVDs | `android avd list` | `maui android emulator list` | ✅ replace |
| Create / start / stop / delete AVD | `android avd …` | `maui android emulator create\|start\|stop\|delete` | ✅ replace |
| Android SDK install / list / info / accept-licenses | `android sdk …` | `maui android sdk …` (+ `maui android install` for guided setup) | ✅ replace |
| JDK install / check | `android jdk …` | `maui android jdk …` | ✅ replace |
| Xcode list | `apple xcode list` | `maui apple xcode list` | ✅ replace |
| Simulator runtime list | (n/a in skills today) | `maui apple runtime list` | ✅ added as a discovery hint |
| List sims | `apple simulator list` / `xcrun simctl list devices` | `maui apple simulator list` | ✅ replace |
| Boot / shutdown / delete sim | `apple simulator boot\|shutdown\|delete` / `xcrun simctl …` | `maui apple simulator start\|stop\|delete` | ✅ replace |
| Environment health check | (n/a) | `maui doctor` | ✅ added |
| Connectivity diagnostic | mixed | `maui devflow diagnose` | ✅ surfaced earlier |
| **Create / erase simulator** | `xcrun simctl create\|erase` | ❌ not in `maui` | ⚠️ kept raw fallback |
| **Install / launch app on sim** | `xcrun simctl install\|launch` | ❌ not in `maui` | ⚠️ kept raw fallback |
| **Sim privacy / appearance / openurl / push / location / addmedia** | `xcrun simctl …` | ❌ not in `maui` | ⚠️ kept raw fallback |
| **Sim screenshot (non-MAUI app)** | `xcrun simctl io booted screenshot` | ❌ not in `maui` (use `maui devflow ui screenshot` for a running MAUI app) | ⚠️ kept raw fallback |
| **Android `adb reverse 19223` (broker)** | `adb reverse tcp:19223 tcp:19223` | ❌ not in `maui` | ⚠️ kept raw; flagged as gap |
| **Android `adb forward <agent port>`** | `adb forward tcp:<port> tcp:<port>` | ❌ not in `maui` | ⚠️ kept raw; flagged as gap |
| **APK install / uninstall / launch / stop** | `adb install` / `am start` / `am force-stop` | ❌ not in `maui` | ⚠️ kept raw |
| **logcat / app logs (pre-agent)** | `adb logcat` | `maui devflow ui logs` works once the agent is up; `adb logcat` still required pre-agent | ⚠️ partial replace |
| **Screencap / screenrecord (Android, pre-agent)** | `adb shell screencap\|screenrecord` | `maui devflow ui screenshot` once agent is up | ⚠️ partial replace |
| **Port-in-use diagnosis** | `lsof -i :<port>` | ❌ not in `maui` | ⚠️ kept raw |
| Build / run app | `dotnet build -t:Run` / `dotnet run` | (intentionally out of `maui` CLI scope) | ✅ keep `dotnet` |

## What changed in the skills (this PR)

All edits are under `plugins/dotnet-maui/skills/`:

- **`maui-devflow-debug/SKILL.md`** — INVOKES line rewritten to lead with the `maui` CLI; new "Reading machine-readable output" section covering `--json`, error code categories (`E1xxx`–`E5xxx`), `remediation.type` (`AutoFixable` / `UserAction` / `Terminal`), and `--ci` for non-interactive failure-fast use.
- **`references/android.md`** — fully rewritten to be `maui`-first. Raw `adb` ops (port forwarding, install, logcat, screencap, push/pull/shell) grouped under a clearly labeled **"Raw fallbacks not yet in `maui` CLI"** section.
- **`references/ios-and-mac.md`** — restructured around `maui apple xcode|runtime|simulator …`. Raw `xcrun simctl` ops (create/erase/install/launch/privacy/appearance/openurl/push/location/addmedia) grouped under the same labeled fallback section. Permission and dark-mode subsections explicitly flag the raw fallback.
- **`references/connectivity.md`** — leads with `maui devflow diagnose` + `maui device list`; new **"Reading errors from the CLI"** section with the structured error envelope and code-prefix table.
- **`references/troubleshooting.md`** — new top section **"Reading machine-readable output and errors"** with the full envelope JSON, code taxonomy, remediation types, and a worked example using `E2106 (AndroidEmulatorNotFound)`. Existing items now reference `maui` commands and concrete codes (`E2101`, `E2103`, `E2402`).
- **`references/setup.md`** — `Microsoft.Maui.Cli` is the only required global tool; `androidsdk.tool` and `appledev.tools` moved to an "Optional extras" subsection. Raw `adb` port-forwarding step labeled. `maui doctor` added to verification.
- **`maui-devflow-onboard/SKILL.md`** — added `maui doctor` to the verify block, plus a one-line note: *"Use `--json` and read `error.code`/`remediation.command` for programmatic recovery."*
- `references/macos.md`, `linux.md`, `batch.md` — verified clean (no changes needed).

### Verification

```
rg -n '^\s*(android |apple |xcrun simctl |adb )' \
  plugins/dotnet-maui/skills/maui-devflow-debug \
  plugins/dotnet-maui/skills/maui-devflow-onboard
```

Every remaining match is inside a clearly labeled "Raw fallbacks not yet in `maui` CLI" section **or** is preceded by a `# Not yet wrapped by 'maui' CLI` comment.

## JSON / error troubleshooting improvements

Skills now teach agents to:

1. Pass `--json` to any command an agent will parse.
2. Branch on `error.code` (categories from `src/Cli/Microsoft.Maui.Cli/Errors/ErrorCodes.cs`):
   - `E1xxx` — tool/internal
   - `E2xxx` — platform/SDK (`E20xx` JDK, `E21xx` Android, `E22xx` Apple, `E23xx` Windows, `E24xx` .NET)
   - `E3xxx` — user action
   - `E4xxx` — network
   - `E5xxx` — permission
3. When `error.remediation.type == AutoFixable`, run `error.remediation.command` and retry.
4. When `error.remediation.type == UserAction`, follow `error.remediation.manualSteps`.
5. Use `--ci` for non-interactive failure-fast runs.

Worked example (E2106) is included in `references/troubleshooting.md`.

## Follow-up backlog — real `maui` CLI gaps

These are out of scope for this PR but should drive future `maui` CLI work so the skills can drop more raw fallbacks:

1. **`maui apple simulator create|erase|install|launch|screenshot|openurl|push|location|addmedia`** — today only `list|start|stop|delete`.
2. **`maui apple simulator privacy grant|revoke|reset`** — for permission management.
3. **`maui apple simulator appearance light|dark`** — for theme toggling in tests.
4. **`maui android port reverse|forward|list|clear`** — wrap `adb reverse`/`adb forward`, with the broker port `19223` as a default and an `--agent-port` shortcut for DevFlow.
5. **`maui android device install|uninstall|launch|stop`** — wrap `adb install -r`, `pm uninstall`, `am start`, `am force-stop`.
6. **`maui android device logcat [--tag …] [--pid …]`** — wrap `adb logcat` with sensible MAUI defaults (`MauiDevFlow`, `DOTNET`).
7. **`maui android device screenshot|screenrecord`** — wrap `adb shell screencap`/`screenrecord`.
8. **`maui port check <port>`** — wrap `lsof -i :<port>` cross-platform (also useful for broker port conflicts).
9. **Structured JSON for errors** — confirm and document that `--json` writes a stable `{ "error": { "code", "message", "remediation": { "type", "command", "manualSteps" }, "context", "nativeError" } }` shape from `MauiToolException`. Add it to skills + CLI README.
10. **`maui devflow diagnose --json`** — ensure the connectivity diagnostic emits a structured per-check result so agents can branch on it instead of grepping prose.

## Out of scope

- The legacy `maui-ai-debugging` skill — it already redirects new clients to `maui-devflow-debug`.
- The `devflow-automation` skill — it's about `[DevFlowAction]`, not device handling.
- No new `maui` CLI commands are added in this PR; only skill markdown is changed.

## References

- Skills: `plugins/dotnet-maui/skills/maui-devflow-debug/`, `plugins/dotnet-maui/skills/maui-devflow-onboard/`
- CLI command surface: `src/Cli/Microsoft.Maui.Cli/Commands/`
- Error taxonomy: `src/Cli/Microsoft.Maui.Cli/Errors/ErrorCodes.cs`
- Error envelope: `src/Cli/Microsoft.Maui.Cli/Errors/MauiToolException.cs`

### PR DESCRIPTION
**Pull request by rmarinho**
Implements `maui port check <port>` from issue #197.

## Summary

Cross-platform TCP port diagnostic tool that identifies which process holds a given port. Useful during MAUI development when DevFlow, emulators, profiling sessions, or other tools need specific ports but get "address already in use" errors.

```
$ maui port check 8080
Port 8080 is in use:
  PID 12345 (dotnet) 0.0.0.0 [ipv4]

$ maui port check 9999
Port 9999 is free.

$ maui port check 8080 --json
{
  "port": 8080,
  "in_use": true,
  "listeners": [
    { "pid": 12345, "process_name": "dotnet", "address": "0.0.0.0:8080", "family": "ipv4", "state": "Listen" }
  ]
}
```

## Platform support

| Platform | Strategy |
|----------|----------|
| **Windows** | P/Invoke `GetExtendedTcpTable` (IPv4 + IPv6), resolves PID and process name natively |
| **macOS/Linux** | Cascading fallback: `lsof` → `ss` → `netstat` → `System.Net.NetworkInformation.IPGlobalProperties` |

The Unix inspector retries process-name resolution with brief back-off to handle race conditions where a process exits between port enumeration and name lookup.

## Exit codes

| Code | Meaning |
|------|---------|
| 0 | Port is free |
| 1 | Port is in use (listeners found) |
| 2 | Error (invalid argument, enumeration failure) |

## Error codes

- **E1008** `PortEnumerationFailed` — raised when the platform inspection fails (e.g., `lsof` not found, P/Invoke error)

## Changes

| File | Purpose |
|------|---------|
| `Commands/PortCommands.cs` | CLI command wiring (`maui port check <port>`) |
| `Providers/Port/IPortInspector.cs` | Platform abstraction interface |
| `Providers/Port/WindowsPortInspector.cs` | Windows implementation via `GetExtendedTcpTable` P/Invoke |
| `Providers/Port/UnixPortInspector.cs` | macOS/Linux with `lsof`/`ss`/`netstat` + managed fallback |
| `Providers/Port/PortListenerInfo.cs` | Shared data model for port listeners |
| `Providers/Port/PortInspectorHelpers.cs` | Process-name resolution with retry logic |
| `Models/PortCheckResult.cs` | JSON-serializable output model |
| `Errors/ErrorCodes.cs` | Added `E1008 PortEnumerationFailed` |
| `MauiCliJsonContext.cs` | AOT-compatible JSON type registration |
| `Program.cs` | Command registration |
| `Tests/PortCheckTests.cs` | 12 unit tests covering parsing, edge cases, and error handling |

## Testing

- 12 unit tests cover: `lsof`/`ss`/`netstat` output parsing, IPv4/IPv6 addresses, listen vs. established state filtering, error fallback paths, invalid port validation
- All tests are cross-platform (parse output strings, no actual port inspection)
- Integration tested manually on macOS and Windows

Closes #197

### CODE REVIEW

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+[SupportedOSPlatform("windows")]
+internal sealed class WindowsPortInspector : IPortInspector
+{
+private const int AF_INET = 2;
+private const int AF_INET6 = 23;
+
+private enum TcpTableClass
+{
+OwnerPidListener = 6,
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed — enum value corrected to `OwnerPidListener = 3`.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+var result = new List<PortListenerInfo>();
+uint bufLen = 0;
+GetExtendedTcpTable(IntPtr.Zero, ref bufLen, false, afFamily, TcpTableClass.OwnerPidListener, 0);
+if (bufLen == 0) return result;
+
+var buffer = Marshal.AllocHGlobal((int)bufLen);
+try
+{
+if (GetExtendedTcpTable(buffer, ref bufLen, false, afFamily, TcpTableClass.OwnerPidListener, 0) != 0)
+return result;
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed — added a retry loop (up to 4 attempts) around `GetExtendedTcpTable` that handles `ERROR_INSUFFICIENT_BUFFER` by doubling the buffer.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+// NAME column may be followed by "(LISTEN)" as a separate token
+var namePart = parts[^1] == "(LISTEN)" && parts.Length >= 10 ? parts[^2] : parts[^1];
+if (!namePart.Contains(':')) continue;
+
+var colonIdx = namePart.LastIndexOf(':');
+var addrRaw = namePart[..colonIdx];
+var portRaw = namePart[(colonIdx + 1)..].TrimEnd(')').Trim();
+if (!int.TryParse(portRaw, out var parsedPort) || parsedPort != port) continue;
+
+var addr = addrRaw == "*" ? "0.0.0.0" : addrRaw;
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Fixed — `ParseLsofOutput` now checks `typeCol == "IPv6"` and maps wildcard to `::` accordingly.

File: src/Cli/Microsoft.Maui.Cli.UnitTests/PortCheckTests.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+var output = "Proto Recv-Q Send-Q Local Address          Foreign Address        (state)\n" +
+             "tcp4  0      0      127.0.0.1.9090         *.*                    LISTEN\n";
+var result = UnixPortInspector.ParseNetstatOutput(output, 8080);
+
+Assert.Empty(result);
+}
+
+[Theory]
+[InlineData(0x5000u, 80)]
+[InlineData(0x1F90u, 8080)]
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli.UnitTests/PortCheckTests.cs]: Fixed — test values corrected to `0x901Fu` for 8080 and `0xBB01u` for 443 (proper network byte order).

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+}
+return result;
+}
+
+internal static int GetPortFromNetworkDword(uint networkDword)
+=> (int)(((networkDword & 0xFF) << 8) | ((networkDword >> 8) & 0xFF));
+
+private static string GetProcessName(int pid)
+{
+try { return Process.GetProcessById(pid).ProcessName; }
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed — `GetProcessName` now wraps in `using var p = Process.GetProcessById(pid)`.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+private static List<PortListenerInfo>? TryLsof(int port)
+{
+try
+{
+var output = RunCommand("lsof", $"-nP -iTCP:{port} -sTCP:LISTEN");
+if (output is null) return null;
+var result = ParseLsofOutput(output, port);
+return result.Count > 0 ? result : null;
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Acknowledged — intentional design choice. The fallthrough cost is negligible for a CLI diagnostic tool (sub-second), and the simpler logic avoids edge cases around distinguishing "tool succeeded with empty output" from "tool not found". Can revisit if needed.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+RedirectStandardOutput = true,
+RedirectStandardError = true,
+UseShellExecute = false,
+CreateNoWindow = true
+}
+};
+process.Start();
+var output = process.StandardOutput.ReadToEnd();
+process.WaitForExit(5000);
+return process.ExitCode == 0 || output.Length > 0 ? output : null;
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Fixed — `RunCommand` now uses async event-based reading (`BeginOutputReadLine`/`BeginErrorReadLine`) to drain both pipes. On timeout, `process.Kill(entireProcessTree: true)` prevents orphans.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+if (ret != ERROR_INSUFFICIENT_BUFFER) return result;
+bufLen = bufLen == 0 ? 8192 : bufLen * 2;
+}
+
+var count = Marshal.ReadInt32(buffer);
+const int rowOffset = 4;
+
+if (afFamily == AF_INET)
+{
+var rowSize = Marshal.SizeOf<MibTcpRowOwnerPid>();
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed in 92f4785 — `ret` is now declared outside the loop and guarded with `if (ret != 0) return result;` before reading the buffer.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+if (!parts[^1].Equals("LISTEN", StringComparison.OrdinalIgnoreCase)) continue;
+
+var localAddr = parts[3];
+var dotIdx = localAddr.LastIndexOf('.');
+if (dotIdx < 0) continue;
+if (!int.TryParse(localAddr[(dotIdx + 1)..], out var parsedPort) || parsedPort != port) continue;
+
+var addrPart = localAddr[..dotIdx];
+var addr = addrPart == "*" ? "0.0.0.0" : addrPart;
+var family = parts[0].Contains('6') ? "ipv6" : "ipv4";
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Fixed in 92f4785 — wildcard now maps to `::` when family is ipv6.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+{
+for (int attempt = 0; attempt < 4; attempt++)
+{
+if (buffer != IntPtr.Zero) Marshal.FreeHGlobal(buffer);
+buffer = Marshal.AllocHGlobal((int)bufLen);
+var ret = GetExtendedTcpTable(buffer, ref bufLen, false, afFamily, TcpTableClass.OwnerPidListener, 0);
+if (ret == 0) break;
+if (ret != ERROR_INSUFFICIENT_BUFFER) return result;
+bufLen = bufLen == 0 ? 8192 : bufLen * 2;
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed in 92f4785 — `ret` is now declared outside the loop and checked after it exits.

[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed in 92f4785 — added `if (bufLen > (uint)int.MaxValue) return result;` guard before AllocHGlobal.

File: src/Cli/Microsoft.Maui.Cli/Commands/PortCommands.cs [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+{
+formatter.WriteSuccess($"Port {port} is free.");
+}
+else
+{
+formatter.WriteInfo($"Port {port} is in use:");
+foreach (var l in result.Listeners)
+formatter.WriteInfo($"  PID {l.Pid} ({l.ProcessName}) {l.Address} [{l.Family}]");
+}
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Commands/PortCommands.cs]: Fixed in 92f4785 — now shows "unknown process" when PID is 0 or process name is empty.

File: src/Cli/Microsoft.Maui.Cli/Commands/PortCommands.cs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+#pragma warning disable CA1416
+inspector = new UnixPortInspector();
+#pragma warning restore CA1416
+}
+}
+catch (Exception ex)
+{
+formatter.WriteError(new MauiToolException(ErrorCodes.PortEnumerationFailed, ex.Message));
+return 2;
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Commands/PortCommands.cs]: The CLI error-reporting convention is to pass `ex.Message` to `MauiToolException` — this is consistent with other commands (AndroidCommands.Jdk.cs, DeviceCommand.cs). The ErrorResult JSON shape surfaces the error code and message; inner exceptions would not serialize cleanly in JSON mode. Verbose/debug logging is not wired into the CLI output path.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+result.Add(new PortListenerInfo(port, 0, string.Empty, addr, family));
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Fixed in 92f4785 — `ParseNetstatOutput` now maps wildcard to `::` when family is ipv6.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Fixed in 92f4785 — `ret` tracked outside loop with guard `if (ret != 0) return result;` before reading buffer.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+
+private static List<PortListenerInfo>? TryNetstat(int port)
+{
+try
+{
+var output = RunCommand("netstat", "-anp tcp");
+if (output is null) return null;
+var result = ParseNetstatOutput(output, port);
+return result.Count > 0 ? result : null;
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Acknowledged — same design choice as finding #7 from the first review round. The fallthrough cost is negligible for a one-shot CLI diagnostic.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+result.AddRange(GetListenersForFamily(port, AF_INET6));
+return result;
+}
+
+private List<PortListenerInfo> GetListenersForFamily(int port, int afFamily)
+{
+var result = new List<PortListenerInfo>();
+uint bufLen = 0;
+GetExtendedTcpTable(IntPtr.Zero, ref bufLen, false, afFamily, TcpTableClass.OwnerPidListener, 0);
+if (bufLen == 0) return result;
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: The initial probe with `bufLen == 0` is expected to return `ERROR_INSUFFICIENT_BUFFER` and populate `bufLen`. If it returns something else and `bufLen` stays 0, the early return on line 65 already handles that. Access-denied errors are not expected for read-only TCP table queries on any supported Windows version.

File: src/Cli/Microsoft.Maui.Cli/Commands/PortCommands.cs [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+namespace Microsoft.Maui.Cli.Commands;
+
+internal static class PortCommands
+{
+public static Command Create()
+{
+var portCommand = new Command("port", "Diagnose TCP port usage.");
+portCommand.Add(CreateCheckCommand());
+return portCommand;
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Commands/PortCommands.cs]: The files use the same indentation style (no tabs) as the original commit. This repo has no .editorconfig and TreatWarningsAsErrors is false, so there is no enforced formatting standard. Happy to reformat if desired, but it would be a style-only change.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+{
+public IReadOnlyList<PortListenerInfo> GetListeners(int port)
+{
+var listeners = TryLsof(port);
+if (listeners is not null) return listeners;
+
+if (OperatingSystem.IsLinux())
+{
+listeners = TrySs(port);
+if (listeners is not null) return listeners;
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/UnixPortInspector.cs]: Same as above — no enforced formatting standard in this repo. Happy to reformat if desired.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+namespace Microsoft.Maui.Cli.Providers.Port;
+
+[SupportedOSPlatform("windows")]
+internal sealed class WindowsPortInspector : IPortInspector
+{
+private const int AF_INET = 2;
+private const int AF_INET6 = 23;
+
+private enum TcpTableClass
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/WindowsPortInspector.cs]: Same as above — no enforced formatting standard in this repo.

File: src/Cli/Microsoft.Maui.Cli/Providers/Port/PortInspectorHelpers.cs [hunk 20]
```diff
@@ -0,0 +1,10 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.Maui.Cli.Providers.Port;
+
+internal static class PortInspectorHelpers
+{
+	internal static int GetPortFromNetworkDword(uint networkDword)
+		=> (int)(((networkDword & 0xFF) << 8) | ((networkDword >> 8) & 0xFF));
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Providers/Port/PortInspectorHelpers.cs:10]: Same as above.

File: src/Cli/Microsoft.Maui.Cli.UnitTests/PortCheckTests.cs [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+
+public class PortCheckTests
+{
+[Fact]
+public void PortCheckResult_JsonShape_HasExpectedProperties()
+{
+var result = new PortCheckResult
+{
+Port = 8080,
+InUse = true,
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli.UnitTests/PortCheckTests.cs]: Same as above.

File: src/Cli/Microsoft.Maui.Cli/Models/PortCheckResult.cs [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+{
+[JsonPropertyName("port")]
+public int Port { get; init; }
+
+[JsonPropertyName("in_use")]
+public bool InUse { get; init; }
+
+[JsonPropertyName("listeners")]
+public List<PortListenerResult> Listeners { get; init; } = [];
+}
```
[the remark this answers is not in the document]
Comment (rmarinho) [reply src/Cli/Microsoft.Maui.Cli/Models/PortCheckResult.cs:18]: Same as above.

**Comment by jfversluis** [comment]
@copilot formatting of this code looks wrong, everything is stuck to the left. Please make it inline with the rest of the codebase.

**Comment by jfversluis** [comment]
The new `maui port check` command is user-facing, but the docs surfaces that list CLI capabilities were not updated. Please add it to the CLI command table in `src/Cli/README.md` and consider adding it to the root README's Cli feature list so users can discover the new diagnostic command.

**Review by jfversluis — APPROVED** [review]
